### PR TITLE
refactor: merging proving module into sales

### LINKED ISSUE #294: Add node restart

### LINKED ISSUE #414: Merge proving and sales state machines
The storage host has two parts: a state machine for setting up the Request (e.g. downloading the info, proving, filling the slot, etc.) which completes with a filled slot, and then the Slot/Request is passed to the proving module. This module handles proving, collecting the reward, and cleaning up. These two parts should be merged.

### PR DESCRIPTION
**Pull request by AuHau**
Closes #414 

Implementation of the simulated proof failure follows a similar path as it was implemented now - a separate state that inherits from the original proving with only `prove` method overridden.

### CODE REVIEW

**Review by emizzle — APPROVED** [review]
Nice work as usual Adam. A few comments, but nothing absolutely blocking.

File: codex/sales/states/filled.nim [hunk 1]
```diff
@@ -21,15 +27,30 @@ method `$`*(state: SaleFilled): string = "SaleFilled"
...
+      onSale(request, slotIndex)
+
+    when codex_enable_proof_failures:
+      if context.simulateProofFailures > 0:
+        info "Proving with failure rate", rate = context.simulateProofFailures
+        return some State(SaleProvingSimulated(failEveryNProofs: context.simulateProofFailures))
+      else:
+        return some State(SaleProving())
+    else:
+      return some State(SaleProving())
```
[the remark this answers is not in the document]
Comment (AuHau) [reply codex/sales/states/filled.nim:52]: We need them.
Line 50 is there for the case when yes the proof-failure compile flag is enabled, but there is actually no runtime CLI/Config flag specified (the `contex.simulateProofFailures` is 0) and hence we won't even turn on the Simulated Proving.

Line 52 is there for the case when no proof-failure compile flag is enabled.

File: codex/sales/states/proving.nim [hunk 2]
```diff
@@ -1,29 +1,77 @@
...
-import ./filled
+import ./errored
+import ./payout
 
 logScope:
     topics = "marketplace sales proving"
 
 type
   SaleProving* = ref object of ErrorHandlingState
+    loop: ?Future[void]
```
[the remark this answers is not in the document]
Comment (AuHau) [reply codex/sales/states/proving.nim:18]: I guess you are right. I just took it over from the current solution without much thinking about it 😅

File: codex/sales/states/proving.nim [hunk 3]
```diff
@@ -1,29 +1,77 @@
...
+
+method prove*(state: SaleProving, slot: Slot, onProve: OnProve, market: Market, currentPeriod: Period) {.base, async.} =
```
[the remark this answers is not in the document]
Comment (AuHau) [reply codex/sales/states/proving.nim]: Sure, just where is this defined?

File: codex/sales/states/proving.nim [hunk 4]
```diff
@@ -1,29 +1,77 @@
...
+    await clock.waitUntil(periodicity.periodStart(period).truncate(int64))
+
+  let slot = Slot(request: request, slotIndex: slotIndex)
+  let slotId = slot.id
+
+  while true:
+    let currentPeriod = await getCurrentPeriod()
+    let slotState = await market.slotState(slot.id)
+    if slotState == SlotState.Finished:
+      debug "Slot reached finished state", period = currentPeriod, requestId = $request.id, slotIndex, slotId = $slot.id
```
[the remark this answers is not in the document]
Comment (AuHau) [reply codex/sales/states/proving.nim:57]: Oh, I did not know about this! Great and thanks!

File: codex/sales/states/proving.nim [hunk 5]
```diff
@@ -1,29 +1,77 @@
...
+
+    await waitUntilPeriod(currentPeriod + 1)
 
 method `$`*(state: SaleProving): string = "SaleProving"
 
 method onCancelled*(state: SaleProving, request: StorageRequest): ?State =
+  if loop =? state.loop:
+      state.loop = Future[void].none
+      if not loop.finished:
+        loop.cancel()
```
[the remark this answers is not in the document]
Comment (AuHau) [reply codex/sales/states/proving.nim:77]: Sounds good! I was not really sure how to go about this as yeah the `onCancelled` is not async, so thanks for great suggestion!

File: codex/sales/states/filled.nim (hunk 1 above)
[the remark this answers is not in the document]
Comment (emizzle) [reply codex/sales/states/filled.nim:52]: Sorry, I didn't explain that correctly. I meant we could simplify it to look like:
```diff
    when codex_enable_proof_failures:
      if context.simulateProofFailures > 0:
        info "Proving with failure rate", rate = context.simulateProofFailures
        return some State(SaleProvingSimulated(failEveryNProofs: context.simulateProofFailures))
-      else:
-        return some State(SaleProving())
-    else:
-      return some State(SaleProving())
+.   return some State(SaleProving())
```

File: codex/sales/states/proving.nim [hunk 6]
```diff
@@ -1,29 +1,106 @@
...
 import ./failed
-import ./filled
+import ./errored
+import ./payout
 
 logScope:
     topics = "marketplace sales proving"
 
 type
+  SlotNotFilled* = object of Exception
```
Comment (emizzle) [inline codex/sales/states/proving.nim:17]: This should be at the very least `object of CatchableError`, but could also be `object of CodexError` (if importing from `../../errors`). `Exception` is used a base for both `CatchableError` and `Defect`, the latter of which is not catchable.

Comment (emizzle) [reply codex/sales/states/proving.nim:17]: Typical nim styling also suffixes with `Error`
```suggestion
  SlotNotFilledError* = object of CatchableError
```

File: codex/sales/states/filled.nim (hunk 1 above)
[the remark this answers is not in the document]
Comment (AuHau) [reply codex/sales/states/filled.nim:52]: Oh right! Nice catch 😅 Thanks!

**Pull request merged**